### ISSUE TITLE
[BUG] cannot send email, private/auth failed: permission denied

### ISSUE THREAD

**Issue by DavyLandman** [docker-mailserver/docker-mailserver#3110]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

smtpd / sasl

### What happened and when does this occur?

Without any updates to the container, user credentials, or vps, I cannot send mail anymore.

I can still connect to the smtp ports (both 465 and 587), it's just that it will get rejected. IMAP receiving of email also still works.

### What did you expect to happen?

I expect the SMTP server to accept my login and allow me to send emails.

Also, it seems to complain about SASL, but that's disabled for my server.

### How do we replicate the issue?

Uhm, I find this a bit hard to say right now, but I would like to help debug and figure out what's going on

### DMS version

v11.3.1

### What operating system is DMS running on?

Linux

### Which operating system version?

Alpine 3.16.4

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker

### docker-compose.yml

```yml
`  mailserver:
    image: docker.io/mailserver/docker-mailserver:11.3.1
    restart: unless-stopped
    stop_grace_period: 1m
    hostname: mail.XXX.cloud
    ports:
      - "25:25"     # SMTP  (explicit TLS => STARTTLS)
      - "143:143"   # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"   # ESMTP (implicit TLS)
      - "587:587"   # ESMTP (explicit TLS => STARTTLS)
      - "993:993"   # IMAP4 (implicit TLS)
      - "4190:4190" # user seive
    volumes:
      - ./data/dms/data:/var/mail
      - ./data/dms/state:/var/mail-state
      - ./logs/dms:/var/log/mail
      # borrow traefik log
      - ./data/traefik/acme.json:/etc/letsencrypt/acme.json:ro
      - /etc/localtime:/etc/localtime:ro
    mem_limit: 3.6G
    cpus: 1.9
    environment:
      - LOG_LEVEL=info
      - SUPERVISOR_LOGLEVEL=info
      - TZ=Europe/Amsterdam
      - POSTFIX_INET_PROTOCOLS=all
      - DOVECOT_INET_PROTOCOLS=all

      - ENABLE_SASLAUTHD=0

      # save all into single dir
      - ONE_DIR=1

      # security/filtering stuff
      # filter stuff
      - SPOOF_PROTECTION=1
      - ENABLE_CLAMAV=1
      - ENABLE_AMAVIS=1
      - ENABLE_POP3=0
      # show warnings from amavis filter
      - AMAVIS_LOGLEVEL=0
      # disable the dns blocklist for now
      - ENABLE_DNSBL=0
      # block attackers
      - ENABLE_FAIL2BAN=1
      - FAIL2BAN_BLOCKTYPE=drop
      # check valid connections with all tests
      - POSTSCREEN_ACTION=enforce
      # enalbe spam checker
      - ENABLE_SPAMASSASSIN=1
      - SPAMASSASSIN_SPAM_TO_INBOX=1
      - ENABLE_SPAMASSASSIN_KAM=1
      - "SA_SPAM_SUBJECT=[SPAM] "
      - ENABLE_MANAGESIEVE=1

      # we'll borrow traefik cretificates
      - SSL_TYPE=letsencrypt
      - SSL_DOMAIN=mail.XXX.cloud

      # limits
      # 50 mb
      - POSTFIX_MESSAGE_SIZE_LIMIT=52428800

    cap_add:
      - NET_ADMIN
      - SYS_PTRACE
    labels:
      - "traefik.enable=false"
```

### Relevant log output

I've looked through the logs (`docker-compose logs --follow`) and noticed this coming up when I tried to send an email (both with thunderbird and with android gmail client).

```Text
mailserver_1                       | Feb 23 19:40:38 mail postfix/submission/smtpd[1455]: warning: hostname XXXXX.connected.by.freedominter.net does not resolve to address 2a10:3781:XXX::1: Name or service not known
mailserver_1                       | Feb 23 19:40:38 mail postfix/submission/smtpd[1455]: connect from unknown[2a10:3781:XXX:1]
mailserver_1                       | Feb 23 19:40:38 mail postfix/submission/smtpd[1455]: Anonymous TLS connection established from unknown[2a10:3781:XXX::1]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (4096 bits) server-digest SHA256
mailserver_1                       | Feb 23 19:40:38 mail postfix/submission/smtpd[1455]: warning: SASL: Connect to private/auth failed: Permission denied
mailserver_1                       | Feb 23 19:40:38 mail postfix/submission/smtpd[1455]: fatal: no SASL authentication mechanisms
mailserver_1                       | Feb 23 19:40:39 mail postfix/master[1399]: warning: process /usr/lib/postfix/sbin/smtpd pid 1455 exit status 1
mailserver_1                       | Feb 23 19:40:39 mail postfix/master[1399]: warning: /usr/lib/postfix/sbin/smtpd: bad command startup -- throttling
```

I've also tried it over ipv4, the same happens.

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am rather experienced with docker
- [X] I am inexperienced with mail servers
- [ ] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [X] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by DavyLandman** [comment]
Hmm:

```
root@mail:/# ls -alh /var/spool/postfix/private/auth
srw-r-----+ 1 docker docker 0 Feb 23 19:53 /var/spool/postfix/private/auth
```

it's s socket, but owned by docker user (uuid 5000) while the rest in that directory is owned by postfix (uid101).

I found some old issues on that file, and someone suggested removing it via `user-patches.sh` so I did that, but that didn't seem to work (and the file is back with the same properties)

I also connected to it:

```
root@mail:/# nc -U /var/spool/postfix/private/auth
VERSION 1       2
MECH    PLAIN   plaintext
MECH    LOGIN   plaintext
SPID    1383
CUID    1
COOKIE  XXXXX
DONE
```

so the socket seems to work. however, if I try it as posfix:

```
root@mail:/# su -l -s /bin/bash postfix
postfix@mail:~$ nc -U /var/spool/postfix/private/auth
nc: unix connect failed: Permission denied
nc: /var/spool/postfix/private/auth: Permission denied
```

so could this be some kind of rights issue? what should I check?

**Comment by DavyLandman** [comment]
I did this:

```
chmod o+rw /var/spool/postfix/private/auth
```

and now it works, but that can't be safe right?

**Comment by georglauterbach** [comment]
This is what the file permissions & group ownership looks like in my container:

```cf
srw-rw-rw- 1 docker docker 0 Feb 23 11:04 /var/spool/postfix/private/auth
```

In my `user-patches.sh`, I run `rm -rf /var/spool/postfix/private/auth`, and it works fine. But I am not sure why I actually need it..

**Comment by georglauterbach** [comment]
So, my research: <https://serverfault.com/questions/227852/what-does-a-mean-at-the-end-of-the-permissions-from-ls-l>. You file has extended permissions (called ACLs). Did you change anything on the host?

**Comment by polarathene** [comment]
This has been reported various times, and I had some detailed comments looking into it, but your report adds some nice insights, thanks :)

This is on my todo list. Last I recall, we have a Dovecot config that specifies this auth file with the docker user, which gets placed in the Postfix location (_prior to v12, this a chroot location and Postfix would duplicate a bunch of files and rewrite permissions for some content, I haven't checked yet if it's the same without chroot in v12, or how that state dir from a prior version would affect anything if the content isn't the same in v12 with fresh state dir_).

From memory, this was configured a long time ago and without much explanation. There's a good chance it was done incorrectly :man_shrugging: It may even be something that worked in the previous Ubuntu base image, but did not with the switch to Debian, or the more recent switch to newer Dovecot?

AFAIK it would only occur with the default `ONE_DIR=1` and the mounted state dir volume. It's most commonly reported from macOS users, where Docker there would be via a interim VM guest running Docker, mounting a host folder that would get sync'd to the VM guest, and then Docker mounting that synced copy. Perhaps some filesystem feature or lock is in play :man_shrugging: 

---

If it really is just a permissions issue, then that should be easier to resolve.

You're both sharing `/var/spool/postfix` as the path, but with the mail state dir bind mounted, start-up scripts should be changing that to a symlink IIRC to `/var/spool-postfix`, and perhaps it's something to do with that which is causing the problem?

---

> and now it works, but that can't be safe right?

:man_shrugging: probably best to consult Dovecot docs on the feature, and try understand why it's added to Postfix path and if it's not been configured correctly in the first place.

**Comment by polarathene** [comment]
> Also, it seems to complain about SASL, but that's disabled for my server.

Incorrect. `ENABLE_SASLAUTHD=0` does not disable SASL, this is actually the default value IIRC.

SASLAUTHD is an alternative SASL backend that was contributed by someone adding LDAP support as they wanted to use Kopano instead of Dovecot for SASL. Our tests additionally cover support for other SASLAUTHD modes beyond LDAP, such as an IMAP based one (rimap) that can be used with Dovecot I think, while the default is to use PAM (_not that anyone uses it or needed that I think_).

If you want to submit mail to be sent from DMS, you need to authenticate, which is delegated to a SASL backend. This is Dovecot by default, if you disable Dovecot with `SMTP_ONLY=1`, then you'll find issues about needing to configure authentication by other means, which AFAIK is handled with SASLAUTHD usually :sweat_smile: 

These are areas of the project that current maintainers haven't worked on much and likely need some improvement / review.

**Comment by DavyLandman** [comment]
> Incorrect. `ENABLE_SASLAUTHD=0` does not disable SASL, this is actually the default value IIRC.

my bad, that was a misfire, seems to not be related to that anyway right?

> Yet AFAIK we don't use docker except for the /var/mail directory which Dovecot manages? So this is probably a misconfiguration tbh.

Kinda strange that the conf said it should be `0666` chmod, while on my machine it got `0600`.

I just removed the file, and restarted the container, and inside the container this is printed:

```
root@mail:/# ls -alh /var/spool/postfix/private/auth
srw-r-----+ 1 docker docker 0 Feb 24 09:22 /var/spool/postfix/private/auth
root@mail:/# getfacl /var/spool/postfix/private/auth
getfacl: Removing leading '/' from absolute path names
# file: var/spool/postfix/private/auth
# owner: docker
# group: docker
user::rw-
user:1001:r-x                   #effective:r--
group::---
mask::r--
other::---
root@mail:/# getfacl -t -e /var/spool/postfix/private/auth
getfacl: Removing leading '/' from absolute path names
# file: var/spool/postfix/private/auth
USER   docker    rw-
user   1001      r-X
GROUP  docker    ---
mask             r--
other            ---
```

So nothing string at facl level, right? 

However, `/var/spool/postfix` is symlinked to `/var/mail-state/spool-postfix` and `/var/mail-state` is mounted outside the container.

if we inspect that file:
```
❯ doas ls -alh data/dms/state/spool-postfix/private/auth
srw-r-----    1 5000     5000           0 Feb 24 09:22 data/dms/state/spool-postfix/private/auth
```

I've also checked facl properties there, and there is one difference:

```
❯ doas getfacl data/dms/state/spool-postfix/private/auth
# file: data/dms/state/spool-postfix/private/auth
# owner: 5000
# group: 5000
user::rw-
user:backup:r-x                 #effective:r--
group::---
mask::r--
other::---

❯ doas getfacl -t -e data/dms/state/spool-postfix/private/auth
# file: data/dms/state/spool-postfix/private/auth
USER   5000      rw-
user   backup    r-X
GROUP  5000      ---
mask             r--
other            ---
```

I've setup a default `facl` for the backup user to be able to read dirs & files in the whole data directory.

So there is some facl involved, but only additive.

**Comment by polarathene** [comment]
> Kinda strange that the conf said it should be `0666` chmod, while on my machine it got `0600`.

Could be due to something happening during start-up?

- Initially the file should not exist, and then Dovecot would start and create it.
- When Postfix is started it would then create `/var/spool/postfix` properly AFAIK (_which perhaps wasn't expecting Dovecot to had preempted it?_).
- Now technically before either starts, we do have start-up script to look at `/var/mail-state` being mounted, and it would then check for the `var/spool/postfix` location (_I don't think this exists on a fresh install on first run?_). There should be no content to operate on, but it should create a symlink anyway.
- Ownership is then changed (_intended to reset to what ownership should be, so this shouldn't do anything.
- When start-up occurs again with existing content. If the `private/auth` file is present in the volume, it'd [become `postfix:root` ownership](https://github.com/docker-mailserver/docker-mailserver/blob/ae05e6a7c379825fb464a648acc04b4074dd547f/target/scripts/startup/misc-stack.sh#L89) at this stage, before Dovecot starts (and presumably corrects it).

That management script for `/var/mail-state` has changed with v12 a bit. And Postfix may behave a bit differently due to no longer running with chroot env, nor delaying starting due removing the wrapper script previous releases had.

That said I think you can change ownership / permissions on symlinks instead of the actual files/directories they link to, so that may also play a role? :sweat_smile: 

I would need to investigate further, but I think we don't need to place this auth socket at that path, it is not something that needs to persist. We should be able to change it at the Dovecot config mentioned earlier, and also in `main.cf`:

https://github.com/docker-mailserver/docker-mailserver/blob/ae05e6a7c379825fb464a648acc04b4074dd547f/target/postfix/main.cf#L75

This could be done with a `user-patches.sh` if you don't want to build an image with modified config. Perhaps that will give better insights since it avoids any race conditions and other changes going on in `/var/mail-state` described above.

**Comment by DavyLandman** [comment]
I think in part it's due to these two relative paths:

https://github.com/docker-mailserver/docker-mailserver/blob/ae05e6a7c379825fb464a648acc04b4074dd547f/target/postfix/master.cf#L22

and

https://github.com/docker-mailserver/docker-mailserver/blob/ae05e6a7c379825fb464a648acc04b4074dd547f/target/postfix/master.cf#L35

if I make those absolute as well, it indeed works.

I just put this in my user-patches.sh as a test:

```
sed -i 's|/var/spool/postfix/private/auth|/var/tmp/sasl-auth|g' /etc/postfix/main.cf /etc/dovecot/conf.d/10-master.conf
sed -i 's|private/auth|/var/tmp/sasl-auth|g' /etc/postfix/master.cf /etc/dovecot/conf.d/10-master.conf
```

so maybe this should be a nice fix around that problem (not my user-patches, but the change of the configs)? to move that socket outside a docker mounted path?

If you want I could try and make a PR for it.

**Comment by georglauterbach** [comment]
> If you want I could try and make a PR for it.

Yes, please go ahead :)

**Comment by DavyLandman** [comment]
Ok, I will, what would the preferred location? Is there a different place
where other temp files are located?

**Comment by georglauterbach** [comment]
I actuall think `/var/tmp/` is a good place :)

**Comment by polarathene** [comment]
> if I make those absolute as well, it indeed works (socket gets right permissions).

That's interesting. I think IIRC, relative paths would usually be from the spool directory in chroot environment, not sure how that differs in v12 since we no longer use that.

I think Dovecot still is responsible for making the socket at the configured path. Then `main.cf` defines that setting with a default value, and for some reason we have that overwritten in master.cf for two smtpd services, possibly from a different contributor if we look at `git blame`, it could just be copy/paste from somewhere else too :man_shrugging: 

**EDIT:** Prior to v12, it's advised to use relative path with Postfix due to chroot. Docs:

- https://doc.dovecot.org/configuration_manual/howto/postfix_and_dovecot_sasl/
- https://www.postfix.org/postconf.5.html#smtpd_sasl_path
- https://www.postfix.org/SASL_README.html#server_dovecot_comm
- https://www.postfix.org/SASL_README.html#server_sasl_enable

Presumably permissions should be for `postfix` not `docker` (Dovecot) too :thinking: 

> what would the preferred location?

It's a unix socket file, so convention is to place that in `/run` I think for system users (Postfix)? So something like: `/run/sasl-auth.sock` maybe? (_`/var/run` symlinks to `/run`_). `/tmp` or `/var/tmp` seems valid too apparently :sweat_smile: 

- https://superuser.com/questions/482977/where-to-put-unix-sockets
- https://unix.stackexchange.com/questions/88083/idiomatic-location-for-file-based-sockets-on-debian-systems
- https://man.archlinux.org/man/file-hierarchy.7

**Comment by georglauterbach** [comment]
I'm more than fine with placing it in `/run/` as `sasl-auth.sock`; seems logical to me.

**Comment by alexanderadam** [comment]
Did I get this right: for now the bugfix is to change the container command to run something like


```bash
sed -i 's|/var/spool/postfix/private/auth|/var/tmp/sasl-auth|g' /etc/postfix/main.cf /etc/dovecot/conf.d/10-master.conf && sed -i 's|private/auth|/var/tmp/sasl-auth|g' /etc/postfix/master.cf && supervisord -c /etc/supervisor/supervisord.conf
```

or is there any better quick fix until the PR is done and released in the container image?

**Comment by polarathene** [comment]
> Did I get this right: for now the bugfix is to change the container command to run something like

You could use our `user-patches.sh` script feature. That'll run a script with the commands (minus the need to start supervisor at the end there), before services like Postfix are started.

All we've discussed is changing the path to be in a separate location which should avoid some potential issues. And the `master.cf` file from relative to absolute path.

However, if you're not running `:edge` / `v12`, then the `master.cf` change might not work :man_shrugging: The linked Dovecot docs mention using a relative path for Postfix chroot (_which is the case for all our current releases, except upcoming v12 that `:edge` provides_), you can try it still and see if that works.

Other than that, the Dovecot config also configured this file to be user and group of `docker` (Dovecot uses this for `/var/mail`), instead of `postfix`. It may not work then if permissions are `660`.

> or is there any better quick fix until the PR is done and released in the container image?

You can technically extend the image with your own `Dockerfile`, and use `COPY` (or `RUN sed` if suitable) to apply such a patch via a custom image. Might be easier if you need to adjust the socket files ownership to be `postfix` instead of `docker`.

**Comment by alexanderadam** [comment]
> You could use our `user-patches.sh` script feature

That makes sense. DMS is still very new to me.

> However, if you're not running `:edge` / `v12`, then the `master.cf` change might not work

It seems you were right with that. I'm using unprivileged Podman with DMS `11.3.1` as this seems to be [the latest stable version](https://hub.docker.com/r/mailserver/docker-mailserver/tags).

```bash
sed -i 's|/var/spool/postfix/private/auth|/var/tmp/sasl-auth|g' /etc/postfix/main.cf /etc/dovecot/conf.d/10-master.conf
sed -i 's|private/auth|/var/tmp/sasl-auth|g' /etc/postfix/master.cf
```

to it and mounted it within the container. I also added some `echo` statements, so that I see that it's executed.
But I'm still getting permission errors:

```bash
  10:23:08 postfix/master[1816]: daemon started -- version 3.5.17, configuration /etc/postfix
  10:23:08,559 INFO success: postfix entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
  10:23:08 postfix/pickup[1819]: fatal: chroot(/var/spool/postfix): Operation not permitted
  10:23:09 postfix/master[1816]: warning: process /usr/lib/postfix/sbin/pickup pid 1819 exit status 1
  10:23:09 postfix/master[1816]: warning: /usr/lib/postfix/sbin/pickup: bad command startup -- throttling
  10:24:09 postfix/pickup[1886]: fatal: chroot(/var/spool/postfix): Operation not permitted
  10:24:10 postfix/master[1816]: warning: process /usr/lib/postfix/sbin/pickup pid 1886 exit status 1
  10:24:10 postfix/master[1816]: warning: /usr/lib/postfix/sbin/pickup: bad command startup -- throttling
```

Everything works on `:edge` though (even without `user-patches.sh`).

Now I'm pretty unsure what I should do as `edge` doesn't sound like something one should use in production.

**Comment by georglauterbach** [comment]
@polarathene @DavyLandman I noticed that our Supervisor socket is located at `/dev/shm/`; do we want the new socket file to be there as well?

**Comment by DavyLandman** [comment]
> > However, if you're not running `:edge` / `v12`, then the `master.cf` change might not work

@polarathene I ran my user-script thingy on 11.3.1 (in Docker). And it's working again after that, the socket gets the right 0666 rights applied, and smtpd works again.

> It seems you were right with that. I'm using unprivileged Podman with DMS `11.3.1` as this seems to be [the latest stable version](https://hub.docker.com/r/mailserver/docker-mailserver/tags).

@alexanderadam can you run some of the commands I ran from within the container? Like, checking the rights of the file and using netcat to read from the socket as different users?

**Comment by georglauterbach** [comment]
ATM there are no major changes to `main.cf`. If you want to give it a try, just go ahead, we'll guide you 🙂

**Comment by alexanderadam** [comment]
> can you run some of the commands I ran from within the container? Like, checking the rights of the file and using netcat to read from the socket as different users?

Sure:

```bash
# ls -alh /var/spool/postfix/private/auth 
srw-rw-rw- 1 docker docker 0 Feb 26 22:52 /var/spool/postfix/private/auth

# getfacl /var/spool/postfix/private/auth
getfacl: Removing leading '/' from absolute path names
# file: var/spool/postfix/private/auth
# owner: docker
# group: docker
user::rw-
group::rw-
other::rw-

# getfacl -t -e /var/spool/postfix/private/auth
getfacl: Removing leading '/' from absolute path names
# file: var/spool/postfix/private/auth
USER   docker    rw-     
GROUP  docker    rw-     
other            rw-

# nc -U /var/spool/postfix/private/auth
VERSION	1	2
MECH	PLAIN	plaintext
MECH	LOGIN	plaintext
SPID	1752
CUID	1
COOKIE	2379c6ff66db5199b0b904d87dfd38ca
DONE

# su -l -s /bin/bash postfix
postfix@host:~$ nc -U /var/spool/postfix/private/auth
VERSION	1	2
MECH	PLAIN	plaintext
MECH	LOGIN	plaintext
SPID	1752
CUID	2
COOKIE	e27635502f150ae0e4a012bf16cb9d33
DONE
```

**Comment by polarathene** [comment]
@alexanderadam

> It seems you were right with that. I'm using unprivileged Podman with DMS `11.3.1`

On `11.3.1`, you will still have Postfix running in a chroot jail, it is restricted to `/var/spool/postfix/` location (_`/var/spool-postfix/` with a `/var/mail-state` volume mounted to the container_).

It doesn't have permission to access files outside of that location AFAIK. We removed this limitation on upcoming v12 release as it doesn't provide much of a security benefit within a Docker container (_already isolates filesystem access_).

The `master.cf` file is using a relative path, which is advised when using the chroot jail.

> But with `:edge` I'm seeing a similar permission error in the log (`setup debug show-mail-logs`) but for Dovecot

Is this `:edge` using the same volume from the `11.3.1` container? Not sure if that'd affect anything, but I am aware that some folders in `/var/mail-state` from previous releases were for the Postfix chroot jail, and no longer relevant, but AFAIK they only log some warnings about that and don't cause any real issues.

Your errors suggest Dovecot is also attempting to use a chroot jail :thinking: 

[`anvil` is part of Postfix](https://www.postfix.org/anvil.8.html), but it also is mentioned in Dovecot docs, and used there apparently. Dovecot also [documents about using chroot too](https://wiki.dovecot.org/Chrooting). I'd probably have to look at our configs to see if there's something there that we need to change too, cheers for sharing that log output :+1: 

---

@alexanderadam I am curious if this is a Podman issue, you're running Podman as root or in rootless mode? Has this file been working without issues for you before as well like it was for @DavyLandman ?

It seems odd that the same changes would work fine for @DavyLandman on the same `11.3.1` release, so I can only assume there is some difference between you two, either in configs or host environment. I'm not sure what caused the permission issue in the first place for a few users that reported in the past either.

The `/var/tmp` location should be fine for the socket with the `docker` user and group Dovecot is configured to create it with, but if Postfix needs it, I assume that's not available to it on `11.3.1`, so added confusion why it's all good now for @DavyLandman :sweat_smile: (_or I am totally mistaken about the chroot jail being relevant for this socket path_)

**Comment by polarathene** [comment]
@georglauterbach 

> I noticed that our Supervisor socket is located at `/dev/shm/`; do we want the new socket file to be there as well?

[`/dev/shm`](https://superuser.com/questions/45342/when-should-i-use-dev-shm-and-when-should-i-use-tmp) didn't seem like a location I'd think storing unix socket files was relevant for. However, I have had a little look into it's relevance with Docker and in particular supervisor to see why we'd have that in the first place...:

- https://github.com/Supervisor/supervisor/issues/654#issuecomment-138369436
- https://github.com/DataDog/docker-dd-agent/pull/269
- https://github.com/docker-mailserver/docker-mailserver/pull/719
- https://github.com/docker-mailserver/docker-mailserver/issues/707

We use it in Docker apparently for ensuring a `tmpfs` mount, avoiding an issue with the default storage driver OverlayFS that didn't play well with this type of file. Looking over the [upstream issue](https://github.com/moby/moby/issues/12080), this was resolved with the 4.7 kernel release (July 2016). I'm skeptical that it's still relevant if it was fixed with a newer kernel. I haven't yet tried running the supervisor socket at it's usual default in `/var/run` instead of `/dev/shm` yet though to confirm.

Perhaps it's possibly related to the problem, and that plays into the different host / Docker environments for users that have always had this issue (_pretty sure I've seen it in the past with Podman and macOS users specifically_).

@alexanderadam you could try seeing if `/dev/shm` works better for you than `/var/tmp`?

**Comment by DavyLandman** [comment]
@alexanderadam but that all looks fine already, so even before this changed it worked okay (at least this socket, you might have a different issue?).

**Comment by alexanderadam** [comment]
> > But with `:edge` I'm seeing a similar permission error in the log (`setup debug show-mail-logs`) but for Dovecot
> 
> Is this `:edge` using the same volume from the `11.3.1` container? Not sure if that'd affect anything, but I am aware that some folders in `/var/mail-state` from previous releases were for the Postfix chroot jail, and no longer relevant, but AFAIK they only log some warnings about that and don't cause any real issues.

This also happens for entirely empty volume mounts.

> Your errors suggest Dovecot is also attempting to use a chroot jail thinking
> 
> [`anvil` is part of Postfix](https://www.postfix.org/anvil.8.html), but it also is mentioned in Dovecot docs, and used there apparently. Dovecot also [documents about using chroot too](https://wiki.dovecot.org/Chrooting). I'd probably have to look at our configs to see if there's something there that we need to change too, cheers for sharing that log output +1

Do you have any clue what I might do for now to circumvent this issue?

> @alexanderadam I am curious if this is a Podman issue, you're running Podman as root or in rootless mode?

It's rootless Podman

> Has this file been working without issues for you before as well like it was for @DavyLandman ?

This is a new installation/

> It seems odd that the same changes would work fine for @DavyLandman on the same `11.3.1` release, so I can only assume there is some difference between you two, either in configs or host environment.

Yes, I guess there are just too many possible causes for permission issues 🫣

> [@alexanderadam](https://github.com/alexanderadam) you could try seeing if `/dev/shm` works better for you than `/var/tmp`?

Could you please elaborate on that?
How can I change that?

**Comment by polarathene** [comment]
> It's rootless Podman

Well rootless + Podman is fairly niche for us. I'd take a guess that Podman without rootless might work, and if so then it's something on your end to resolve regarding rootless support.

You'll find in our docs there have been some contributions from users of Podman and rootless. I'm not sure if any of that will help.

My usual approach is to get a working setup going, then gradually adjust to the broken setup. My bet is on rootless being the problem though, so I'd suggest looking into that first. If it is, then you'd need to try figure out if it's possible to resolve for rootless, it would likely be a common painpoint with rootless (possibly affects Docker rootless in the same way). Perhaps there are options to configure the rootless mode, or some well documented gotchas to be aware of, I don't use rootless myself.

---

> Could you please elaborate on that?
> How can I change that?

You had some `sed` commands that were changing paths to `/var/tmp`, I was suggesting trying to change to `/dev/shm` instead, as there is a small chance of that making a difference.

---

> I think everything runs fine and I just can't connect because of a port forwarding issue — which isn't related to DMS of course.
> Maybe this message is just timing related or so?

It could be a variety of things for sure. Unfortunately it's not something any maintainers are able to support beyond our knowledge of the project. We have regular Docker and k8s experience, none of us run rootless or with Podman where the problem seems to be most common with the few reports we have received.

You could try running an alternative mail project that has Postfix + Dovecot. You may run into the same issues, or they have different config/approach that resolves it (potentially something we can reference and consider), or their own docs might better communicate what needs to be done for a rootless + Podman environment.
